rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+- Start Date: 2022-06-13
+- RFC PR: [#63](https://github.com/inveniosoftware/rfcs/pull/63)
+- Authors: David Eckhard, Zacharias Zacharodimos, Alex Ioannidis
+
+# Notification System
+
+## Summary
+
+Implementation details for a notification system to be used in InvenioRDM.
+
+## Motivation
+
+As the product is getting larger and new features are added, it is hard to keep an overview of everything going on as a user. With the communities feature, certain tasks require an action to be executed by the user (f.e. accepting an incoming request for a community). This incoming request can easily get "lost", if a user does not explicitly check the community dashboard. Therefore, a notification system will not only solve this problem (and similar ones) but also increase user engagement by advertising the repository some more.
+
+It will take care of sending the notifications based on the configuration and user preferences (i.e. if the repository disables certain notifications or a user does not want to receive certain notifications, they should not be sent).
+
+### Use Case
+
+- Send notifications for operations involving multiple users (see [table](#notification-examples))
+- Overridable notification templates
+  - One template for each notification
+  - Can be overridden via theme (like other templates)
+  - Add new templates for custom notifications
+  - Should be translatable
+    - Store preferred language in user profile
+- Modify notifications
+  - Set which notifications should be send
+    - As a repository admin activate/deactive certain notifications
+    - As a user (user profile)
+  - Define notifications for events per module
+  - Define notification backends per module
```

Comment:
How is it going to look for like a user? 

I'd like to see
- Profile page: Disable/enable notification backends, edit preferred language
- Examples of possible notifications sent by e.g. email and say Mattermost/Slack? E.g. the email should likely have information about how to "unsubscribe" - i.e. change their notification settings. On Zenodo we may want to ask users wan to receive informational emails so we're able to send messages to all users that want to receive them.

Let me know if you'd like me to help with mockups

Reply:
After chat, what is probably missing here is to describe with more details that the templates will be also per-backend.
And yes, the plan was to have e-mail notifications, and immediately add `CERN Mattermost` as extra to test that it actually works adding another backend.

Reply:
> How is it going to look for like a user?

Mockups at https://github.com/inveniosoftware/invenio-communities/issues/864#issuecomment-1497107633. But not part of the initial implementation because of time-constraints and planning requirements. Will add to the RFC though for reference.

We're still missing mockups for the rendered email/chat templates, in the current implementation we'll just have the (hyper)text message. These can be mocked-up and styled later on as we did e.g. for registration, and email confirmation emails.

> Question: If we wanted to build a page listing notifications for a user, would this module be able to support it? i..e would it just be a matter of a new backend storing the notifications?

Yes, we can just add a backend that is used to persist the message to primary/secondary storage. Probably this can live in this module and have it's own service, resources, UI in the user settings page, etc.